lib/blockstore/file.js

```diff
@@ -0,0 +1,789 @@
+   * @returns {Promise}
+   */
+
+  async writeMerkle(hash, data) {
+    return this._write(types.MERKLE, hash, data);
+  }
+
+  /**
+   * This method stores block undo coin data in files.
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async writeUndo(hash, data) {
+    return this._write(types.UNDO, hash, data);
+  }
+
+  /**
+   * This method stores block data in files.
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async writeBlock(hash, data) {
+    return this._write(types.BLOCK, hash, data);
+  }
+
+  /**
+   * This method stores block data in files with by appending
+   * data to the last written file and updating indexes to point
+   * to the file and position.
+   * @private
+   * @param {Number} type - The type of block data
+   * @param {Buffer} hash - The block hash
+   * @param {Buffer} data - The block data
+   * @returns {Promise}
+   */
+
+  async _write(type, hash, data) {
+    if (this.writing[type])
+      throw new Error('Already writing.');
+
+    this.writing[type] = true;
+
+    if (await this.db.has(layout.b.encode(type, hash))) {
+      this.writing[type] = false;
+      return false;
+    }
+
+    let mlength = 8;
+
+    // Hash for a block is not stored with
+    // the magic prefix as it's read from the header
+    // of the block data.
+    if (type !== types.BLOCK && type !== types.MERKLE)
+      mlength += 32;
+
+    const blength = data.length;
+    const length = data.length + mlength;
+
+    const bwm = bio.write(mlength);
+
+    bwm.writeU32(this.network.magic);
+    bwm.writeU32(blength);
+
+    if (type !== types.BLOCK && type !== types.MERKLE)
+      bwm.writeHash(hash);
+
+    const magic = bwm.render();
+
+    const {
+      fileno,
+      filerecord,
+      filepath
+    } = await this.allocate(type, length);
+
+    const mposition = filerecord.used;
+    const bposition = filerecord.used + mlength;
+
+    const fd = await fs.open(filepath, 'r+');
+
+    let mwritten = 0;
+    let bwritten = 0;
+
+    try {
+      mwritten = await fs.write(fd, magic, 0, mlength, mposition);
+      bwritten = await fs.write(fd, data, 0, blength, bposition);
+    } finally {
+      await fs.close(fd);
+    }
+
+    if (mwritten !== mlength) {
+      this.writing[type] = false;
+      throw new Error('Could not write block magic.');
+    }
+
+    if (bwritten !== blength) {
+      this.writing[type] = false;
+      throw new Error('Could not write block.');
+    }
+
+    filerecord.blocks += 1;
+    filerecord.used += length;
+
+    const b = this.db.batch();
+
+    const blockrecord = new BlockRecord({
+      file: fileno,
+      position: bposition,
+      length: blength
+    });
+
+    b.put(layout.b.encode(type, hash), blockrecord.encode());
+    b.put(layout.f.encode(type, fileno), filerecord.encode());
+
+    const last = bio.write(4).writeU32(fileno).render();
```

Comment:
Could this be optimized with `Buffer.alloc(4).writeUInt32BE(fileno)` ? I suppose we assume that `bio` has already allocated memory and the `render()` function has negligible overhead?

Reply:
yep, `bio.write(size)` with size will use StaticWriter which allocates all at once instead of BufferWriter which is dynamic. If you don't pass `size` then you end up with BufferWriter and yeah render just returns allocated buffer (only checks for offset for out of bounds write)